🎉 New source: Fauna

macmv opened the pull request:
## What
Add [Fauna Source](https://fauna.com).


## 🚨 User Impact 🚨
This is a new source, so no user impact is expected.

## Pre-merge Checklist

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [x] Documentation updated
    - [x] Connector's `README.md`
    - [x] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [x] `docs/integrations/README.md`
    - [x] `airbyte-integrations/builds.md`
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

## Tests

<details><summary><strong>Unit</strong></summary>

```
~/workplace/fauna-airbte/airbyte/airbyte-integrations/connectors/source-fauna (source-fauna) » python3.9 -m pytest                                                                                 cleve@Cleves-MBP
Test session starts (platform: darwin, Python 3.9.7, pytest 6.2.5, pytest-sugar 0.9.4)
cachedir: .pytest_cache
rootdir: /Users/cleve/workplace/fauna-airbte/airbyte, configfile: pytest.ini
plugins: sugar-0.9.4, timeout-1.4.2
collecting ... Expr({'now': None})
Expr({'paginate': Expr({'documents': Expr({'collection': 'foo'})}), 'size': 1})
Expr({'exists': Expr({'index': 'ts'})})
Expr({'select': 'source', 'from': Expr({'get': Expr({'index': 'ts'})})})
Expr({'select': 'values', 'from': Expr({'get': Expr({'index': 'ts'})}), 'default': Expr([])})
status=<Status.SUCCEEDED: 'SUCCEEDED'> message=None

 airbyte-integrations/connectors/source-fauna/unit_tests/check_test.py::test_valid_query ✓                                                                                                             4% ▌         
<MagicMock id='4486726992'>                                                                                                                                                                                        

 airbyte-integrations/connectors/source-fauna/unit_tests/check_test.py::test_invalid_check ✓                                                                                                           9% ▉         
Expr({'now': None})                                                                                                                                                                                                
Expr({'paginate': Expr({'documents': Expr({'collection': 'foo'})}), 'size': 1})
None
Expr({'now': None})
Expr({'paginate': Expr({'documents': Expr({'collection': 'invalid_collection_name'})}), 'size': 1})
Collection 'invalid_collection_name' does not exist
Expr({'now': None})
Expr({'paginate': Expr({'documents': Expr({'collection': 'foo'})}), 'size': 1})
Expr({'exists': Expr({'index': 'ts'})})
Expr({'select': 'source', 'from': Expr({'get': Expr({'index': 'ts'})})})
Expr({'select': 'values', 'from': Expr({'get': Expr({'index': 'ts'})}), 'default': Expr([])})
None
Expr({'now': None})
Expr({'paginate': Expr({'documents': Expr({'collection': 'foo'})}), 'size': 1})
Expr({'exists': Expr({'index': 'invalid_index_name'})})
Index 'invalid_index_name' does not exist
Expr({'now': None})
Expr({'paginate': Expr({'documents': Expr({'collection': 'foo'})}), 'size': 1})
Expr({'exists': Expr({'index': 'wrong_source'})})
An exception occurred: invalid query Expr({'exists': Expr({'index': 'wrong_source'})})
Expr({'now': None})
Expr({'paginate': Expr({'documents': Expr({'collection': 'foo'})}), 'size': 1})
Expr({'exists': Expr({'index': 'extra_values_index'})})
Expr({'select': 'source', 'from': Expr({'get': Expr({'index': 'extra_values_index'})})})
Expr({'select': 'values', 'from': Expr({'get': Expr({'index': 'extra_values_index'})}), 'default': Expr([])})
None
Expr({'now': None})
Expr({'paginate': Expr({'documents': Expr({'collection': 'foo'})}), 'size': 1})
Expr({'exists': Expr({'index': 'no_values_index'})})
Expr({'select': 'source', 'from': Expr({'get': Expr({'index': 'no_values_index'})})})
Expr({'select': 'values', 'from': Expr({'get': Expr({'index': 'no_values_index'})}), 'default': Expr([])})
Index should have values [{'field': 'ts'}, {'field': 'ref'}], but it has values []

 airbyte-integrations/connectors/source-fauna/unit_tests/check_test.py::test_check_fails ✓                                                                                                            13% █▍        
Expr({'now': None})                                                                                                                                                                                                
Expr({'paginate': Expr({'documents': Expr({'collection': 'foo'})}), 'size': 1})
Expr({'exists': Expr({'index': 'ts'})})
Expr({'select': 'source', 'from': Expr({'get': Expr({'index': 'ts'})})})
Expr({'select': 'values', 'from': Expr({'get': Expr({'index': 'ts'})}), 'default': Expr([])})
Expr({'now': None})
Expr({'paginate': Expr({'documents': Expr({'collection': 'foo'})}), 'size': 1})
Expr({'exists': Expr({'index': 'ts'})})
Expr({'select': 'source', 'from': Expr({'get': Expr({'index': 'ts'})})})
Expr({'select': 'values', 'from': Expr({'get': Expr({'index': 'ts'})}), 'default': Expr([])})

 airbyte-integrations/connectors/source-fauna/unit_tests/check_test.py::test_config_columns ✓                                                                                                         17% █▊        
Waiting for FaunaDB to start...                                                                                                                                                                                    
.............
FaunaDB is ready! Starting tests
Setting up database...
Database is setup!
[info]: reading add/removes for stream foo
[info]: emitting object {'ref': '105', 'ts': 1659566139180000, 'my_deletion_col': '2022-08-03T22:35:39.180000'}
Ref(id=339012264410481152, collection=Ref(id=foo, collection=Ref(id=collections))) Ref(id=339012264456618496, collection=Ref(id=foo, collection=Ref(id=collections))) Ref(id=339012264498561536, collection=Ref(id=foo, collection=Ref(id=collections)))
[info]: reading add/removes for stream foo
[info]: emitting object {'ref': '339012264410481152', 'ts': 1659566139650000, 'my_deletion_col': '2022-08-03T22:35:39.650000'}
[info]: emitting object {'ref': '339012264498561536', 'ts': 1659566139690000, 'my_deletion_col': '2022-08-03T22:35:39.690000'}
[info]: emitting object {'ref': '339012264456618496', 'ts': 1659566139730000, 'my_deletion_col': '2022-08-03T22:35:39.730000'}
=== check: make sure we read the initial state
[info]: state: {}
[info]: syncing stream 'foo' with incremental
[info]: skipping collection events (no deletions needed)
[info]: reading document updates for stream foo
=== check: make sure the state resumes
[info]: state: {'foo': {'updates_cursor': {'ts': 1659566138700000, 'ref': '104'}}}
[info]: syncing stream 'foo' with incremental
[info]: skipping collection events (no deletions needed)
[info]: reading document updates for stream foo
=== check: make sure that updates are actually read
[info]: state: {'foo': {'updates_cursor': {'ts': 1659566138700000, 'ref': '104'}}}
[info]: syncing stream 'foo' with incremental
[info]: skipping collection events (no deletions needed)
[info]: reading document updates for stream foo
=== check: make sure we read the initial state
[info]: state: {}
[info]: syncing stream 'deletions_test' with incremental
[info]: reading add/removes for stream deletions_test
[info]: reading document updates for stream deletions_test
=== check: make sure we don't produce more records when nothing changed
[info]: state: {'deletions_test': {'remove_cursor': {'ts': 0, 'ref': ''}, 'updates_cursor': {'ts': 1659566140490000, 'ref': '104'}}}
[info]: syncing stream 'deletions_test' with incremental
[info]: reading add/removes for stream deletions_test
[info]: reading document updates for stream deletions_test
=== check: make sure deleted documents produce records
[info]: state: {'deletions_test': {'remove_cursor': {'ts': 0, 'ref': ''}, 'updates_cursor': {'ts': 1659566140490000, 'ref': '104'}}}
[info]: syncing stream 'deletions_test' with incremental
[info]: reading add/removes for stream deletions_test
[info]: emitting object {'ref': '101', 'ts': 1659566140820000, 'deleted_at': '2022-08-03T22:35:40.820000'}
[info]: emitting object {'ref': '103', 'ts': 1659566140850000, 'deleted_at': '2022-08-03T22:35:40.850000'}
[info]: reading document updates for stream deletions_test
=== check: make sure we don't produce more deleted documents when nothing changed
[info]: state: {'deletions_test': {'remove_cursor': {'ts': 1659566140850000, 'ref': '103'}, 'updates_cursor': {'ts': 1659566140490000, 'ref': '104'}}}
[info]: syncing stream 'deletions_test' with incremental
[info]: reading add/removes for stream deletions_test
[info]: reading document updates for stream deletions_test
=== check: make sure another deleted document produces one more record
[info]: state: {'deletions_test': {'remove_cursor': {'ts': 1659566140850000, 'ref': '103'}, 'updates_cursor': {'ts': 1659566140490000, 'ref': '104'}}}
[info]: syncing stream 'deletions_test' with incremental
[info]: reading add/removes for stream deletions_test
[info]: emitting object {'ref': '102', 'ts': 1659566141020000, 'deleted_at': '2022-08-03T22:35:41.020000'}
[info]: reading document updates for stream deletions_test
=== check: make sure we don't produce more deleted documents when nothing changed
[info]: state: {'deletions_test': {'remove_cursor': {'ts': 1659566141020000, 'ref': '102'}, 'updates_cursor': {'ts': 1659566140490000, 'ref': '104'}}}
[info]: syncing stream 'deletions_test' with incremental
[info]: reading add/removes for stream deletions_test
[info]: reading document updates for stream deletions_test
Stopping FaunaDB container...
Stopped FaunaDB container

 airbyte-integrations/connectors/source-fauna/unit_tests/database_test.py::test_incremental_reads ✓                                                                                                   22% ██▎       
 airbyte-integrations/connectors/source-fauna/unit_tests/discover_test.py::test_simple_discover ✓                                                                                                     26% ██▋       
 airbyte-integrations/connectors/source-fauna/unit_tests/discover_test.py::test_discover_valid_index ✓                                                                                                30% ███▏      
 airbyte-integrations/connectors/source-fauna/unit_tests/discover_test.py::test_config_columns ✓                                                                                                      35% ███▌      
 airbyte-integrations/connectors/source-fauna/unit_tests/discover_test.py::test_discover_extra_columns ✓                                                                                              39% ███▉      
[info]: no after token, starting from beginning                                                                                                                                                                    

 airbyte-integrations/connectors/source-fauna/unit_tests/full_refresh_test.py::test_read_all ✓                                                                                                        43% ████▍     
[info]: no after token, starting from beginning                                                                                                                                                                    

 airbyte-integrations/connectors/source-fauna/unit_tests/full_refresh_test.py::test_read_all_extra_columns ✓                                                                                          48% ████▊     
[info]: no after token, starting from beginning                                                                                                                                                                    
[info]: using after token ['even more magical', 3, 'data']

 airbyte-integrations/connectors/source-fauna/unit_tests/full_refresh_test.py::test_read_all_resume ✓                                                                                                 52% █████▎    
[info]: state: {}                                                                                                                                                                                                  
[info]: syncing stream 'my_stream_name' with incremental

 airbyte-integrations/connectors/source-fauna/unit_tests/incremental_test.py::test_read_no_updates_or_creates_but_removes_present ✓                                                                   57% █████▋    
[info]: state: {}                                                                                                                                                                                                  
[info]: syncing stream 'my_stream_name' with incremental
[info]: skipping collection events (no deletions needed)

 airbyte-integrations/connectors/source-fauna/unit_tests/incremental_test.py::test_read_updates_ignore_deletes ✓                                                                                      61% ██████▏   
[info]: reading add/removes for stream foo                                                                                                                                                                         
[info]: emitting object {'ref': '100', 'ts': 99, 'deletes_here': '1970-01-01T00:00:00.000099'}
[info]: emitting object {'ref': '5', 'ts': 999, 'deletes_here': '1970-01-01T00:00:00.000999'}
[info]: reading add/removes for stream foo
[info]: emitting object {'ref': '3', 'ts': 12345, 'deletes_here': '1970-01-01T00:00:00.012345'}
[info]: reading add/removes for stream foo

 airbyte-integrations/connectors/source-fauna/unit_tests/incremental_test.py::test_read_removes_resume_from_partial_failure ✓                                                                         65% ██████▌   
[info]: reading add/removes for stream foo                                                                                                                                                                         
[info]: emitting object {'ref': '100', 'ts': 1648944000000000.0, 'my_deleted_column': '2022-04-03T00:00:00'}
[info]: reading add/removes for stream foo
[info]: reading add/removes for stream foo
[info]: emitting object {'ref': '300', 'ts': 1648944001000000.0, 'my_deleted_column': '2022-04-03T00:00:01'}

 airbyte-integrations/connectors/source-fauna/unit_tests/incremental_test.py::test_read_remove_deletions ✓                                                                                            70% ██████▉   
[info]: reading document updates for stream my_stream_name                                                                                                                                                         
[info]: reading document updates for stream my_stream_name
[info]: reading document updates for stream my_stream_name

 airbyte-integrations/connectors/source-fauna/unit_tests/incremental_test.py::test_read_updates_query ✓                                                                                               74% ███████▍  
[info]: reading document updates for stream my_stream_name                                                                                                                                                         
[info]: reading document updates for stream my_stream_name

 airbyte-integrations/connectors/source-fauna/unit_tests/incremental_test.py::test_read_updates_resume ✓                                                                                              78% ███████▉  
 airbyte-integrations/connectors/source-fauna/unit_tests/serialize_test.py::test_date ✓                                                                                                               83% ████████▍ 
 airbyte-integrations/connectors/source-fauna/unit_tests/serialize_test.py::test_fauna_time ✓                                                                                                         87% ████████▊ 
 airbyte-integrations/connectors/source-fauna/unit_tests/serialize_test.py::test_bytes ✓                                                                                                              91% █████████▎
 airbyte-integrations/connectors/source-fauna/unit_tests/serialize_test.py::test_ref ✓                                                                                                                96% █████████▋
 airbyte-integrations/connectors/source-fauna/unit_tests/serialize_test.py::test_recursive ✓                                                                                                         100% ██████████
================================================================================================= warnings summary =================================================================================================
../../../../../../.asdf/installs/python/3.9.7/lib/python3.9/site-packages/hyper/http11/connection.py:13
../../../../../../.asdf/installs/python/3.9.7/lib/python3.9/site-packages/hyper/http11/connection.py:13
  /Users/cleve/.asdf/installs/python/3.9.7/lib/python3.9/site-packages/hyper/http11/connection.py:13: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3, and in 3.10 it will stop working
    from collections import Iterable, Mapping

../../../../../../.asdf/installs/python/3.9.7/lib/python3.9/site-packages/hyperframe/flags.py:14
  /Users/cleve/.asdf/installs/python/3.9.7/lib/python3.9/site-packages/hyperframe/flags.py:14: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3, and in 3.10 it will stop working
    class Flags(collections.MutableSet):

../../../../../../.asdf/installs/python/3.9.7/lib/python3.9/site-packages/hyper/common/headers.py:13
  /Users/cleve/.asdf/installs/python/3.9.7/lib/python3.9/site-packages/hyper/common/headers.py:13: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3, and in 3.10 it will stop working
    class HTTPHeaderMap(collections.MutableMapping):

-- Docs: https://docs.pytest.org/en/stable/warnings.html

Results (21.32s):
      23 passed
```

</details>

<details><summary><strong>Integration</strong></summary>

`unit_tests/database_test.py` tests exporting data from a local Fauna container. These tests pass, and are grouped in with the unit tests above.

</details>

<details><summary><strong>Acceptance</strong></summary>

```
~/workplace/fauna-airbte/airbyte/airbyte-integrations/connectors/source-fauna (source-fauna) » python3.9 -m pytest -p integration_tests.acceptance
Test session starts (platform: darwin, Python 3.9.7, pytest 6.2.5, pytest-sugar 0.9.4)
cachedir: .pytest_cache
rootdir: /Users/cleve/workplace/fauna-airbte/airbyte, configfile: pytest.ini
plugins: sugar-0.9.4, timeout-1.4.2
collecting ... 
 ../../fauna-airbyte-source/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_config_match_spec[inputs0] ✓3% ▍                                                                                                
 ../../fauna-airbyte-source/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_match_expected[inputs0] ✓6% ▋                                                                                                   
 ../../fauna-airbyte-source/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_docker_env[inputs0] ✓8% ▉                                                                                                       
 ../../fauna-airbyte-source/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_oneof_usage[inputs0] ✓11% █▎                                                                                                    
 ../../fauna-airbyte-source/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_required[inputs0] ✓14% █▍                                                                                                       
 ../../fauna-airbyte-source/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_optional[inputs0] ✓17% █▋                                                                                                       
 ../../fauna-airbyte-source/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_has_secret[inputs0] ✓19% █▉                                                                                                     
 ../../fauna-airbyte-source/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_secret_never_in_the_output[inputs0] ✓22% ██▎                                                                                    
 ../../fauna-airbyte-source/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_defined_refs_exist_in_json_spec_file[inputs0] ✓25% ██▌                                                                          
 ../../fauna-airbyte-source/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_oauth_flow_parameters[inputs0] ✓28% ██▊                                                                                         
 ../../fauna-airbyte-source/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestConnection.test_check[inputs0] ✓31% ███▏                                                                                                  
 ../../fauna-airbyte-source/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestConnection.test_check[inputs1] ✓33% ███▍                                                                                                  
 ../../fauna-airbyte-source/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestConnection.test_check[inputs2] ✓36% ███▋                                                         ../../fauna-airbyte-source/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_discover[inputs0] ✓                                                39% ███▉      
 ../../fauna-airbyte-source/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_discover[inputs1] ✓                                       42% ████▎     
 ../../fauna-airbyte-source/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_defined_cursors_exist_in_schema[inputs0] ✓                44% ████▌     
 ../../fauna-airbyte-source/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_defined_cursors_exist_in_schema[inputs1] ✓                47% ████▊     
 ../../fauna-airbyte-source/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_defined_refs_exist_in_schema[inputs0] ✓                   50% █████     
 ../../fauna-airbyte-source/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_defined_refs_exist_in_schema[inputs1] ✓                   53% █████▍    
 ../../fauna-airbyte-source/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_defined_keyword_exist_in_schema[inputs0-allOf] ✓          56% █████▋    
 ../../fauna-airbyte-source/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_defined_keyword_exist_in_schema[inputs0-not] ✓            58% █████▉    
 ../../fauna-airbyte-source/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_defined_keyword_exist_in_schema[inputs1-allOf] ✓          61% ██████▎   
 ../../fauna-airbyte-source/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_defined_keyword_exist_in_schema[inputs1-not] ✓            64% ██████▍   
 ../../fauna-airbyte-source/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_primary_keys_exist_in_schema[inputs0] ✓                   67% ██████▋   
 ../../fauna-airbyte-source/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_primary_keys_exist_in_schema[inputs1] ✓                   69% ██████▉   
{"type": "LOG", "log": {"level": "INFO", "message": "Actual records for stream sample-data:"}}                                                                                                                     
{"type": "LOG", "log": {"level": "INFO", "message": "[\n {\n  \"ref\": \"337567897171787849\",\n  \"ts\": 1658188683585000,\n  \"data\": {\n   \"a\": 5,\n   \"nested\": {\n    \"value\": 15\n   }\n  },\n  \"my_a_col\": 5,\n  \"nested_column\": 15,\n  \"optional_name\": null\n },\n {\n  \"ref\": \"337567897172836425\",\n  \"ts\": 1658271973660000,\n  \"data\": {\n   \"a\": 6,\n   \"nested\": {\n    \"value\": 20\n   },\n   \"name\": \"hello world\"\n  },\n  \"my_a_col\": 6,\n  \"nested_column\": 20,\n  \"optional_name\": \"hello world\"\n },\n {\n  \"ref\": \"337567897172837449\",\n  \"ts\": 1658188683585000,\n  \"data\": {\n   \"a\": 7,\n   \"nested\": {\n    \"value\": 25\n   }\n  },\n  \"my_a_col\": 7,\n  \"nested_column\": 25,\n  \"optional_name\": null\n },\n {\n  \"ref\": \"337567897173885001\",\n  \"ts\": 1658188683585000,\n  \"data\": {\n   \"a\": 8,\n   \"nested\": {\n    \"value\": 30\n   }\n  },\n  \"my_a_col\": 8,\n  \"nested_column\": 30,\n  \"optional_name\": null\n }\n]"}}
{"type": "LOG", "log": {"level": "INFO", "message": "Expected records for stream sample-data:"}}
{"type": "LOG", "log": {"level": "INFO", "message": "[\n {\n  \"ref\": \"337567897171787849\",\n  \"ts\": 1658188683585000,\n  \"data\": {\n   \"a\": 5,\n   \"nested\": {\n    \"value\": 15\n   }\n  },\n  \"my_a_col\": 5,\n  \"nested_column\": 15,\n  \"optional_name\": null\n },\n {\n  \"ref\": \"337567897172836425\",\n  \"ts\": 1658271973660000,\n  \"data\": {\n   \"a\": 6,\n   \"nested\": {\n    \"value\": 20\n   },\n   \"name\": \"hello world\"\n  },\n  \"my_a_col\": 6,\n  \"nested_column\": 20,\n  \"optional_name\": \"hello world\"\n },\n {\n  \"ref\": \"337567897172837449\",\n  \"ts\": 1658188683585000,\n  \"data\": {\n   \"a\": 7,\n   \"nested\": {\n    \"value\": 25\n   }\n  },\n  \"my_a_col\": 7,\n  \"nested_column\": 25,\n  \"optional_name\": null\n },\n {\n  \"ref\": \"337567897173885001\",\n  \"ts\": 1658188683585000,\n  \"data\": {\n   \"a\": 8,\n   \"nested\": {\n    \"value\": 30\n   }\n  },\n  \"my_a_col\": 8,\n  \"nested_column\": 30,\n  \"optional_name\": null\n }\n]"}}

 ../../fauna-airbyte-source/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestBasicRead.test_read[inputs0] ✓                                           72% ███████▎  
{"type": "LOG", "log": {"level": "INFO", "message": "Actual records for stream deletions-data:"}}                                                                                                                  
{"type": "LOG", "log": {"level": "INFO", "message": "[\n {\n  \"ref\": \"338836293305763911\",\n  \"ts\": 1659398359360000,\n  \"deleted_at\": \"2022-08-01T23:59:19.360000\"\n },\n {\n  \"ref\": \"338836293305761863\",\n  \"ts\": 1659398366330000,\n  \"deleted_at\": \"2022-08-01T23:59:26.330000\"\n },\n {\n  \"ref\": \"338836293305765959\",\n  \"ts\": 1659398371330000,\n  \"deleted_at\": \"2022-08-01T23:59:31.330000\"\n },\n {\n  \"ref\": \"338836293305762887\",\n  \"ts\": 1659398320430000,\n  \"data\": {\n   \"a\": 6,\n   \"nested\": {\n    \"value\": 20\n   }\n  }\n },\n {\n  \"ref\": \"338836293305764935\",\n  \"ts\": 1659398320430000,\n  \"data\": {\n   \"a\": 8,\n   \"nested\": {\n    \"value\": 30\n   }\n  }\n }\n]"}}
{"type": "LOG", "log": {"level": "INFO", "message": "Expected records for stream deletions-data:"}}
{"type": "LOG", "log": {"level": "INFO", "message": "[\n {\n  \"ref\": \"338836293305763911\",\n  \"ts\": 1659398359360000,\n  \"deleted_at\": \"2022-08-01T23:59:19.360000\"\n },\n {\n  \"ref\": \"338836293305761863\",\n  \"ts\": 1659398366330000,\n  \"deleted_at\": \"2022-08-01T23:59:26.330000\"\n },\n {\n  \"ref\": \"338836293305765959\",\n  \"ts\": 1659398371330000,\n  \"deleted_at\": \"2022-08-01T23:59:31.330000\"\n },\n {\n  \"ref\": \"338836293305762887\",\n  \"ts\": 1659398320430000,\n  \"data\": {\n   \"a\": 6,\n   \"nested\": {\n    \"value\": 20\n   }\n  }\n },\n {\n  \"ref\": \"338836293305764935\",\n  \"ts\": 1659398320430000,\n  \"data\": {\n   \"a\": 8,\n   \"nested\": {\n    \"value\": 30\n   }\n  }\n }\n]"}}

 ../../fauna-airbyte-source/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestBasicRead.test_read[inputs1] ✓                                           75% ███████▌  
{"type": "LOG", "log": {"level": "ERROR", "message": "Docker container was failed, code 1, error:\n{\"type\": \"TRACE\", \"trace\": {\"type\": \"ERROR\", \"emitted_at\": 1659565939374.643, \"error\": {\"message\": \"Something went wrong in the connector. See the logs for more details.\", \"internal_message\": \"2 validation errors for ConfiguredAirbyteCatalog\\nstreams -> 0 -> sync_mode\\n  value is not a valid enumeration member; permitted: 'full_refresh', 'incremental' (type=type_error.enum; enum_values=[<SyncMode.full_refresh: 'full_refresh'>, <SyncMode.incremental: 'incremental'>])\\nstreams -> 0 -> destination_sync_mode\\n  value is not a valid enumeration member; permitted: 'append', 'overwrite', 'append_dedup' (type=type_error.enum; enum_values=[<DestinationSyncMode.append: 'append'>, <DestinationSyncMode.overwrite: 'overwrite'>, <DestinationSyncMode.append_dedup: 'append_dedup'>])\", \"stack_trace\": \"Traceback (most recent call last):\\n  File \\\"/airbyte/integration_code/main.py\\\", line 13, in <module>\\n    launch(source, sys.argv[1:])\\n  File \\\"/usr/local/lib/python3.9/site-packages/airbyte_cdk/entrypoint.py\\\", line 118, in launch\\n    for message in source_entrypoint.run(parsed_args):\\n  File \\\"/usr/local/lib/python3.9/site-packages/airbyte_cdk/entrypoint.py\\\", line 106, in run\\n    config_catalog = self.source.read_catalog(parsed_args.catalog)\\n  File \\\"/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/source.py\\\", line 28, in read_catalog\\n    return ConfiguredAirbyteCatalog.parse_obj(self.read_config(catalog_path))\\n  File \\\"pydantic/main.py\\\", line 521, in pydantic.main.BaseModel.parse_obj\\n  File \\\"pydantic/main.py\\\", line 341, in pydantic.main.BaseModel.__init__\\npydantic.error_wrappers.ValidationError: 2 validation errors for ConfiguredAirbyteCatalog\\nstreams -> 0 -> sync_mode\\n  value is not a valid enumeration member; permitted: 'full_refresh', 'incremental' (type=type_error.enum; enum_values=[<SyncMode.full_refresh: 'full_refresh'>, <SyncMode.incremental: 'incremental'>])\\nstreams -> 0 -> destination_sync_mode\\n  value is not a valid enumeration member; permitted: 'append', 'overwrite', 'append_dedup' (type=type_error.enum; enum_values=[<DestinationSyncMode.append: 'append'>, <DestinationSyncMode.overwrite: 'overwrite'>, <DestinationSyncMode.append_dedup: 'append_dedup'>])\\n\", \"failure_type\": \"system_error\"}}}\n"}}

 ../../fauna-airbyte-source/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestBasicRead.test_airbyte_trace_message_on_failure[inputs0] ✓               78% ███████▊  
{"type": "LOG", "log": {"level": "ERROR", "message": "Docker container was failed, code 1, error:\n{\"type\": \"TRACE\", \"trace\": {\"type\": \"ERROR\", \"emitted_at\": 1659565941167.9648, \"error\": {\"message\": \"Something went wrong in the connector. See the logs for more details.\", \"internal_message\": \"2 validation errors for ConfiguredAirbyteCatalog\\nstreams -> 0 -> sync_mode\\n  value is not a valid enumeration member; permitted: 'full_refresh', 'incremental' (type=type_error.enum; enum_values=[<SyncMode.full_refresh: 'full_refresh'>, <SyncMode.incremental: 'incremental'>])\\nstreams -> 0 -> destination_sync_mode\\n  value is not a valid enumeration member; permitted: 'append', 'overwrite', 'append_dedup' (type=type_error.enum; enum_values=[<DestinationSyncMode.append: 'append'>, <DestinationSyncMode.overwrite: 'overwrite'>, <DestinationSyncMode.append_dedup: 'append_dedup'>])\", \"stack_trace\": \"Traceback (most recent call last):\\n  File \\\"/airbyte/integration_code/main.py\\\", line 13, in <module>\\n    launch(source, sys.argv[1:])\\n  File \\\"/usr/local/lib/python3.9/site-packages/airbyte_cdk/entrypoint.py\\\", line 118, in launch\\n    for message in source_entrypoint.run(parsed_args):\\n  File \\\"/usr/local/lib/python3.9/site-packages/airbyte_cdk/entrypoint.py\\\", line 106, in run\\n    config_catalog = self.source.read_catalog(parsed_args.catalog)\\n  File \\\"/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/source.py\\\", line 28, in read_catalog\\n    return ConfiguredAirbyteCatalog.parse_obj(self.read_config(catalog_path))\\n  File \\\"pydantic/main.py\\\", line 521, in pydantic.main.BaseModel.parse_obj\\n  File \\\"pydantic/main.py\\\", line 341, in pydantic.main.BaseModel.__init__\\npydantic.error_wrappers.ValidationError: 2 validation errors for ConfiguredAirbyteCatalog\\nstreams -> 0 -> sync_mode\\n  value is not a valid enumeration member; permitted: 'full_refresh', 'incremental' (type=type_error.enum; enum_values=[<SyncMode.full_refresh: 'full_refresh'>, <SyncMode.incremental: 'incremental'>])\\nstreams -> 0 -> destination_sync_mode\\n  value is not a valid enumeration member; permitted: 'append', 'overwrite', 'append_dedup' (type=type_error.enum; enum_values=[<DestinationSyncMode.append: 'append'>, <DestinationSyncMode.overwrite: 'overwrite'>, <DestinationSyncMode.append_dedup: 'append_dedup'>])\\n\", \"failure_type\": \"system_error\"}}}\n"}}

 ../../fauna-airbyte-source/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestBasicRead.test_airbyte_trace_message_on_failure[inputs1] ✓               81% ████████▏ 
 ../../fauna-airbyte-source/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_full_refresh.py::TestFullRefresh.test_sequential_reads[inputs0] ✓                     83% ████████▍ 
 ../../fauna-airbyte-source/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_incremental.py::TestIncremental.test_two_sequential_reads[inputs0] ✓                  86% ████████▋ 
 ../../fauna-airbyte-source/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_incremental.py::TestIncremental.test_two_sequential_reads[inputs1] ✓                  89% ████████▉ 
 ../../fauna-airbyte-source/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_incremental.py::TestIncremental.test_read_sequential_slices[inputs0] ✓                92% █████████▎
 ../../fauna-airbyte-source/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_incremental.py::TestIncremental.test_read_sequential_slices[inputs1] ✓                94% █████████▌
 ../../fauna-airbyte-source/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_incremental.py::TestIncremental.test_state_with_abnormally_large_values[inputs0] ✓    97% █████████▊
 ../../fauna-airbyte-source/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_incremental.py::TestIncremental.test_state_with_abnormally_large_values[inputs1] ✓   100% ██████████
{"type": "LOG", "log": {"level": "INFO", "message": "/Users/cleve/workplace/fauna-airbte/airbyte/airbyte-integrations/connectors/source-fauna - SAT run - d9f845ec899f964faeba4d7791b59d05ec98f736 - PASSED"}}


Results (49.18s):
      36 passed
```

</details>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/airbytehq/airbyte?pullRequest=15274) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/airbytehq/airbyte?pullRequest=15274) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/airbytehq/airbyte?pullRequest=15274) it.</sub>

marcosmarxm commented:
@macmv sorry the long delay to review the contribution!

marcosmarxm commented on airbyte-integrations/connectors/source-fauna/source_fauna/spec.yaml:
```diff
@@ -0,0 +1,210 @@
+      title: Fauna Secret
+      description: Fauna secret, used when authenticating with the database.
+      airbyte_secret: true
+    collection:
```
Every time a user must pass all collection they want to sync with every field and data type?

marcosmarxm replied:
This must be executed in Airbyte code, remove it from the user responsability.

cleve-fauna replied:
No, this is not how this works.

Fauna stores documents. By default all fields in the user's document will be exported.

However, if the user wishes to elevate certain _sub-fields_ of their document to top-level columns in the destination they can do so with these controls.

For collections, yes user's will have to setup a connector per collection. This is by design. As it enables users to specify different subfield controls as well as makes incremental sync a simple setup for the user (due to details of Fauna's internals).

Both of these choices were intentional and should give the user's a better experience.

marcosmarxm replied:
> However, if the user wishes to elevate certain sub-fields of their document to top-level columns in the destination they can do so with these controls.

For that the user can use dbt in destination to change data whenever they want. Following all other databases the user should input user/password/host/port and Airbyte connector will shows all table/collection/documents to user in the Connection Setup step. The user will have the option to select/unselect which one they don't want to sync.

I'll take a deeper look how incremental will work with Fauna. I'd just setup a test account 💯 

@misteryeo could you give your thoughts here?

cleve-fauna replied:
This does not preclude the usage of `dbt`. We thought it would be nice to include this for user's so they could set this once and forget it without having to build _any_ code at all.

Requiring them to use `dbt` forces them to maintain their own software.

With this we give user's a zero-touch option; which is what our customer's tell us they like about Fauna.

grishick replied:
I am going so second what @marcosmarxm said earlier. The need to specify collection details during connector setup makes this DB connector behavior different from all other DB connectors and also makes it harder to a user to sync multiple collections. Ideally, the connector will implement `discover` method that would return all collections and then user can chose which collections to sync.

On the second part of the comment about selecting certain fields and elevating certain sub-fields - this is a neat feature and we would eventually want to have it for all DB connectors. We plan to add the ability to select which fields to sync at the platform level (https://github.com/airbytehq/airbyte/issues/2227). And we also plan to make normalization more flexible. However, if you keep the current implementation in this connector, it will not be compatible with this feature in the future and will become hard to maintain. So, my suggestion is to not offer this feature at this point and to make the Fauna source connector behave more like other Database connectors. Specifically:

- let the user see all available collections and select the ones that the user wants to sync
- let the user see all available fields and add the code, which will read the field list from the platform as part of catalog. This way, when field filtering feature is implemented, this connector will get it for free

macmv replied:
We're experimenting with this internally.

One thing to note is that fauna has no schema, so the best we can do is provide a stream which just has a couple fields:
- `ref` The document's unique ID
- `ts` The timestamp of the most recent modification to the document
- `ttl` The time to live for the document
- `data` A JSON blob containing all other data for the document.

This format is difficult to work with in the destination, so it will require users to use DBT. If you'd like us to avoid the additionalColumns configuration, then this is the best we can do.

Additionally, this will not be fixed by #2227, as that only looks like that will allow field filtering. We don't have a list of fields at all, so this won't change things. Also, having the additionalColumns config will still be compatible with #2227, as you will still need that configuration in order for the source to know about the columns in the first place.

macmv replied:
How do other sources handle this? There are a number of schemaless databases, and the sources need to produce some type of schema for these databases.

Any recommendations on how we should produce a schema would be appreciated! Having a single `data` field would be clunky, and that seems like a last resort to me.

grishick replied:
Generally, in Airbyte's framework, sources do not care about destinations and vice versa. Right now, we do not have a good solution for normalizing schemaless data at destinations, especially if a destination is a SQL database, not a document store. There are two options here:

1. Read a sample of documents and report a set of fields in those documents as the schema. This is what MongoDB source does (see the code [here](https://github.com/airbytehq/airbyte/blob/f5bfe031eb04aa81614a34382e9c72a2899c840d/airbyte-db/db-lib/src/main/java/io/airbyte/db/mongodb/MongoUtils.java#L203))
2. Do what you are suggesting (return `ref` as a PK, `ts` as a cursor, `data` as a column that contains JSON representation of all fields)

First solution would make it easier for users who have similar documents. Second solution would be more flexible, but will require that users write additional normalization steps, which is not something we allow right now in the Cloud, but is on our roadmap. Do you know what is the more prevalent use pattern for Fauna customers?

macmv replied:
We've updated the connector to export four columns: data, ref, ts, and ttl. These will contain all the data for the Fauna document.

Sorry this took so long, we needed to have some internal discussion about how to proceed for this one.

grishick commented on airbyte-integrations/connectors/source-fauna/examples/README.md:
````diff
@@ -0,0 +1,39 @@
+
+Once started the container is up, setup the database:
+```
+fauna eval "$(cat examples/setup_database.fql)" --domain localhost --port 8443 --scheme http --secret secret
````
It would be worth adding instructions on how to install fauna-shell (https://docs.fauna.com/fauna/current/build/integrations/shell/). I tried following these instructions and had to spend at least 10 minutes trying to figure out where that `fauna` CLI comes from.

macmv replied:
The most recent instructions can be found on the fauna shell github repo: https://github.com/fauna/fauna-shell

I've added a note about that to the README in the examples.

grishick reviewed (changes requested):
I added a couple of comments inline. 

Additionally, please fix the issues reported by 
```
./gradlew :airbyte-integrations:connectors:source-fauna:flakeCheck 
```

Otherwise `build` task fails:

```
./gradlew :airbyte-integrations:connectors:source-fauna:build 
```

grishick commented on airbyte-integrations/connectors/source-fauna/examples/config_sample.json:
```diff
@@ -0,0 +1,16 @@
+{
+  "secret": "SECRET",
+  "domain": "db.fauna.com",
```
I tried this with a test database that I created on https://dashboard.fauna.com/, but it did not work. The error I get is `Failed to connect to database: Unauthorized`. I am using the secret key from https://dashboard.fauna.com.

Given that Fauna has a cloud based service, it would make sense to have an easy to follow example for setting up a connection to cloud hosted Fauna database.

macmv replied:
Did you set the `domain` field correctly? It needs to be set based on the region you chose.

See the region group docs: https://docs.fauna.com/fauna/current/learn/understanding/region_groups#how-to-use-region-groups

grishick replied:
Thanks for the pointer! I did not set the domain correctly as it turns out. I couldn't find those details on my Fauna cloud dashboard. I will retest

macmv replied:
Yeah, we're going to fix that Soon™ :)

I'm glad it works! Let me know how your test goes.

macmv commented:
This latest commit should fix all the warnings produce by running
```
./gradlew :airbyte-integrations:connectors:source-fauna:flakeCheck 
```

macmv commented:
Hi @grishick, I was wondering if you could take a look at this PR sometime soon?

grishick commented:
Looks great! I think there are several small things left to tidy up:

1. @marcosmarxm can you make sure that the credentials for integration tests are in LastPass and are named in accordance with what the README.md says
2. @macmv in the README.md in addition to saying "create a file `secrets/config.json` conforming to the `source_fauna/spec.yaml` file" please add an example contents of `config.json` file or mention, which example file in examples folder to use as a base
3. examples/secret_config.json still has `additional_columns`, which, as far as I can see, have been removed from `spec.yaml` - need to bring this up to date as well
4. rebase and update the pr

grishick commented:
@macmv could you ping me on Airbyte Slack and share `config.json` file that I can use to run integration tests?

macmv commented:
1. @marcosmarxm do you also want me to send you the `secrets.json` file?
2. The `examples/secret_config.json` is a good example of `secrets/config.json`, and I've added that to the README.
3. Yes, that config was out of date. I've updated that as well.
4. Rebased.

I just sent you a message on Slack, which contains the `secrets.json` file.

marcosmarxm commented:
Thanks for the contribution @macmv

cleve-fauna commented:
Thank you @marcosmarxm and @grishick !
